### PR TITLE
Change in  io_handler::get_output_directory to improve output directories

### PR DESCRIPTION
**Pull request by GernotMaier**
This changes allows to separate the directories:

- with results from a simtool
- temporary / intermediate files like model files, ray-tracing files, etc.

This is especially important for using simtools with the output_path and the use_plain_output_path option. In this case we are writing:
- result to `output_path`
- e.g. model files to `output_path/model`.
This should allow the CWL workflows to find the output files.

This pull request required a lot of repetitive changes (you know I like sed). I suggest to first look at simtools/io_handler.py and maybe start with tests/unit_tests/test_io_handler.py::test_get_output_directory. The unit test should cover all the logically allowed cases.

Note that some files also experienced a lot of re-formatting due to possible first-time application of black in the pre-commit stage. 

(and I know that the name of the branch is not good).

### CODE REVIEW

**Review by orelgueta — COMMENTED** [review]
I will do a full review tomorrow, but from a quick look into `io_handler` I had a basic question already.

File: simtools/io_handler.py [hunk 1]
```diff
@@ -85,16 +87,14 @@ def get_output_directory(self, label=None, dir_type=None, test=False):
...
             path = Path(self.output_path)
         else:
-            if test:
-                output_directory_prefix = Path(self.output_path).joinpath("test-output")
-            else:
-                output_directory_prefix = Path(self.output_path).joinpath("simtools-output")
-
-            today = datetime.date.today()
-            label_dir = label if label is not None else "d-" + str(today)
+            output_directory_prefix = Path(self.output_path).joinpath(dir_type[:8] + "-output")
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py:90]: This construct allows to handle the special case of `dir_type = simtools-results` without too many if / else statements. Any better idea to do that efficiently?

What is missing is to handle the 'None' case, otherwise there is no danger in terms of length of arrays:

```
>>> dir_type="a"
>>> print(dir_type[:8])
a
>>> dir_type=None
>>> print(dir_type[:8])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'NoneType' object is not subscriptable
```

**Review by VictorBarbosaMartins — COMMENTED** [review]
Hi @GernotMaier. Thanks for these changes and for spotting the missing pre-commit from my side. I left some comments to improve the clarity of the code/doc.

I did not see how these changes are going to help us distinguish the output of the application/results from the all the other temporary files (as we discussed). It seems to me that all the files (and outputs) will land inside Path(f"{args_dict['output_path']}/output/model") (I did not have time yet to test the applications to confirm this). I think by addressing/responding to my comments, this will become clearer for me. Thank you.
(For the record this review took me 1h30min.)

**Comment by GernotMaier** [comment]
> Hi @GernotMaier. Thanks for these changes and for spotting the missing pre-commit from my side. I left some comments to improve the clarity of the code/doc.
> 
> I did not see how these changes are going to help us distinguish the output of the application/results from the all the other temporary files (as we discussed). It seems to me that all the files (and outputs) will land inside Path(f"{args_dict['output_path']}/output/model") (I did not have time yet to test the applications to confirm this). I think by addressing/responding to my comments, this will become clearer for me. Thank you. (For the record this review took me 1h30min.)

@VictorBarbosaMartins - thanks a lot the review!

I am trying to understand your comment above. Maybe it would have been better not to review and ask immediately  for clarification. This has been extensively tested and it provides a separation of the output files as outlined in the introduction to the pull request.

To make it clearer (as discussed also on Tuesday in our meeting):

1. The situation before this pull request is that we have a inconsistent path selection for output among the tools: this is hard to handle, as we need to know in which directories are the output files.

e.g., running  `simtools-derive-mirror-rnda` with `--output_path ./output` and `--use_plain_output_path` gives:
- derived model parameters are on `./output/results`
- downloaded model files from DB are in `./output/`
- temporary files like photons from ray-tracing are in `./output`

2. With this pull request we always have
- derive model parameter in `./output`
- downloaded model files from DB are in `./output/model/`
- temporary files like photons from ray-tracing are in `./output/raytracing`

This has been tested for several tools and is also consistently tested in the unit test.

I have put a lot of care in making sure that the default behavior of the tools is not changing. Meaning without `--output_path` and `--use_plain_output_path` it is the simtools-output/..date.../...tool..name etc output. We could discuss if this is needed, but this is not the goal of this pull request.

For using `def get_output_directory(self, label=None, sub_dir=None, dir_type="simtools"):`

- nothing changed for the general case (giving label and sub_dir only
- for testing, the boolean test is replaced by the feature parameter "test"
- new: to force the output consistent with the schema outlined in 2. above: use the directory type "simtools-output" when getting/setting the directory of a simtools.

File: simtools/io_handler.py [hunk 2]
```diff
@@ -59,18 +59,20 @@ def set_paths(
         self.data_path = data_path
         self.model_path = model_path
 
-    def get_output_directory(self, label=None, dir_type=None, test=False):
+    def get_output_directory(self, label=None, sub_dir=None, dir_type="simtools"):
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py:62]: Yes, this is a `directory_type` (see explanation below).

We are using both `dir` and `directory` throughout the code, so this is consistent with our naming. I suggest to open an issue to discuss if we should change this.

File: simtools/io_handler.py [hunk 3]
```diff
@@ -59,18 +59,20 @@ def set_paths(
...
         ----------
         label: str
             Instance label.
-        dir_type: str
+        sub_dir: str
             Name of the subdirectory (ray-tracing, model etc)
-        test: bool
-            If true, return test output location
+        dir_type: str
+            Directory type (e.g., 'simtools', 'test', 'simtools-result')
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py:73]: So this is a general question: can we assume that `e.g.` is understood as for example? We have been using it a lot to indicate that these are examples and in my opinion this is sufficiently clear. But given your comment, we should re-discuss this and if possible change everywhere (turn this into an issue?).

File: simtools/io_handler.py [hunk 4]
```diff
@@ -59,18 +59,20 @@ def set_paths(
...
+            Using 'simtools-result' will return the output directory without
+            subdirectory string added for the use_plain_output_path option.
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py:75]: I reformulated it. Let me know if this is ok.

File: simtools/io_handler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py]: Could you please indicate what the actual issue is? Not liking it is not enough for me to work with.

[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py]: First comment: this ensures that the simtools behave as they did before and we have a consistent naming to earlier versions of the code.

Second comment: this is naming which existing in the code and is not connected to this pull request. I suggest either an issue or a separate pull request if there are good reasons to change it.

File: simtools/io_handler.py [hunk 5]
```diff
@@ -85,16 +87,14 @@ def get_output_directory(self, label=None, dir_type=None, test=False):
...
-            label_dir = label if label is not None else "d-" + str(today)
+            output_directory_prefix = Path(self.output_path).joinpath(dir_type[:8] + "-output")
+            label_dir = label if label is not None else "d-" + str(datetime.date.today())
             path = output_directory_prefix.joinpath(label_dir)
-            if dir_type is not None:
-                path = path.joinpath(dir_type)
+        if sub_dir is not None:
+            if not self.use_plain_output_path:
+                path = path.joinpath(sub_dir)
+            elif dir_type != "simtools-result":
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py:100]: Good point, but actually not needed here. For the `else` case, `path` is set to the correct value already in the lines before these if statements.

File: simtools/io_handler.py [hunk 6]
```diff
@@ -114,17 +114,17 @@ def get_output_file(self, file_name, label=None, dir_type=None, test=False):
             File name.
         label: str
             Instance label.
-        dir_type: str
+        sub_dir: str
             Name of the subdirectory (ray-tracing, model etc)
-        test: bool
-            If true, return test output location
+        dir_type: str
+            Directory type (e.g., 'simtools', 'test', 'simtools-result')
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py:124]: Added the improved version here.

File: simtools/io_handler.py [hunk 7]
```diff
@@ -114,17 +114,17 @@ def get_output_file(self, file_name, label=None, dir_type=None, test=False):
...
 
         Returns
         -------
         Path
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py:128]: Let's separate this - there are many things to improve in the code and we should not water down this pull request by adding additional items. I would prefer to have changes reviewed.

File: simtools/camera_efficiency.py [hunk 8]
```diff
@@ -60,8 +60,8 @@ def __init__(
         self.io_handler = io_handler.IOHandler()
         self._base_directory = self.io_handler.get_output_directory(
             label=self.label,
-            dir_type="camera-efficiency",
-            test=test,
+            sub_dir="camera-efficiency",
+            dir_type="test" if self.test else "simtools",
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/camera_efficiency.py:64]: docstring has been improved, so this should be covered.

File: simtools/data_model/model_data_writer.py [hunk 9]
```diff
@@ -28,7 +28,9 @@ def __init__(self, product_data_file=None, product_data_format=None):
         self._logger = logging.getLogger(__name__)
         self.io_handler = io_handler.IOHandler()
         try:
-            self.product_data_file = self.io_handler.get_output_file(file_name=product_data_file)
+            self.product_data_file = self.io_handler.get_output_file(
+                file_name=product_data_file, dir_type="simtools_result"
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/data_model/model_data_writer.py:32]: This is a typo and it should be "simtools-result". Fixed it and also checked for other occurrences (this is the only one)

File: simtools/io_handler.py [hunk 10]
```diff
@@ -85,16 +87,18 @@ def get_output_directory(self, label=None, dir_type=None, test=False):
...
-                output_directory_prefix = Path(self.output_path).joinpath("test-output")
-            else:
-                output_directory_prefix = Path(self.output_path).joinpath("simtools-output")
-
-            today = datetime.date.today()
-            label_dir = label if label is not None else "d-" + str(today)
+            try:
+                output_directory_prefix = Path(self.output_path).joinpath(dir_type[:8] + "-output")
+            except TypeError:
+                self._logger.error(f"Error creating output directory name from {self.dir_type}")
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py:93]: I actually don't expect users to run into this errors (as dir_type is not a user configurable parameter). This is an error message for developers, and there we can expect that they check the API of the method and fix the error.

File: tests/unit_tests/test_io_handler.py [hunk 11]
```diff
@@ -8,68 +8,91 @@
 
 
 def test_get_output_directory(args_dict, io_handler):
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply tests/unit_tests/test_io_handler.py:10]: OK.

File: tests/unit_tests/test_io_handler.py [hunk 12]
```diff
@@ -8,68 +8,91 @@
...
         f"{args_dict['output_path']}/output/test-output/test-io-handler/"
     )
 
-    assert io_handler.get_output_directory(label="test-io-handler", dir_type="model") == Path(
+    # label and subdirectory (no testing)
+    assert io_handler.get_output_directory(label="test-io-handler", sub_dir="model") == Path(
         f"{args_dict['output_path']}/output/simtools-output/test-io-handler/model"
     )
 
+    # label and subdirectory (no testing); simtools-results should have no effect
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply tests/unit_tests/test_io_handler.py:26]: Disagree. Parameters are often not on their own but depend on other settings. So in this case it is important that it doesn't have an effect.

File: tests/unit_tests/test_io_handler.py [hunk 13]
```diff
@@ -8,68 +8,91 @@
...
+    assert io_handler.get_output_directory(
+        label="test-io-handler", sub_dir="model", dir_type="simtools-results"
+    ) == Path(f"{args_dict['output_path']}/output/simtools-output/test-io-handler/model")
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply tests/unit_tests/test_io_handler.py:29]: This is not new and is consistent with the current implementation. I was trying not to change the behaviour of the simtools for the default settings.

File: tests/unit_tests/test_io_handler.py [hunk 14]
```diff
@@ -8,68 +8,91 @@
...
+
+    # label and subdirectory (testing)
     assert io_handler.get_output_directory(
-        label="test-io-handler", dir_type="model", test=True
+        label="test-io-handler", sub_dir="model", dir_type="test"
     ) == Path(f"{args_dict['output_path']}/output/test-output/test-io-handler/model")
 
+    # all following tests: plain_path tests
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply tests/unit_tests/test_io_handler.py:36]: Not entirely sure what you mean with that. Could you explain a bit more what warning you expect?

File: tests/unit_tests/test_io_handler.py [hunk 15]
```diff
@@ -8,68 +8,91 @@
...
+    # (no testing)
+    assert io_handler.get_output_directory(label="test-io-handler", sub_dir="model") == Path(
+        f"{args_dict['output_path']}/output/model"
     )
 
+    # plain path, label has no effect, with sub directory as dir_type != 'simtools-results
+    # (testing)
+    assert io_handler.get_output_directory(
+        label="test-io-handler", sub_dir="model", dir_type="test"
+    ) == Path(f"{args_dict['output_path']}/output/model")
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply tests/unit_tests/test_io_handler.py:59]: Same comment as before: has been in simtools since ever and the intention of the pull request is not to change the existing default behavior.

File: simtools/io_handler.py [hunk 16]
```diff
@@ -85,16 +87,18 @@ def get_output_directory(self, label=None, dir_type=None, test=False):
...
         else:
-            if test:
-                output_directory_prefix = Path(self.output_path).joinpath("test-output")
-            else:
-                output_directory_prefix = Path(self.output_path).joinpath("simtools-output")
-
-            today = datetime.date.today()
-            label_dir = label if label is not None else "d-" + str(today)
+            try:
+                output_directory_prefix = Path(self.output_path).joinpath(dir_type[:8] + "-output")
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py:91]: Good, thanks again for all of your comments. I think the only one related to this pull request is the first one, where I want to get feedback from both of you if the current implementation is ok. It is valid and efficient python code, so please let me know what the actual issue is.

All other points are not really related to this pull request (naming is consistent with the rest of the code; output directories are written the same way as before).

**Comment by GernotMaier** [comment]
I think I've addressed the comments with the last commit. 

Let me know about remaining issues.

File: simtools/io_handler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/io_handler.py:62]: Ok, I opened the issue for that.

File: simtools/io_handler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/io_handler.py]: I think using 'e.g.' as 'example' is fine. My suggestion would be to constrain the choices for this parameter to a selection of expected (acceptable) names. Any arguments against constraining/limiting the choices?

File: simtools/io_handler.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/io_handler.py]: I am still missing explanations for type. I am not sure whether this would be clear for others but I would not know what to choose as `dir_type`. Is the choice really supposed to be free?

My general feeling is that we don't need a lot of options. Maybe the three examples you gave are sufficient? 

Wouldn't it be better to use the same flag in case of use_plain_output_path, i.e., instead of `simtools-result` we use `use_plain_output_path` for dir_type?

File: simtools/io_handler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/io_handler.py]: I thought that was clear, but I can rephrase it. Since I can pass any name in `dir_type` I like, I could choose e.g. 'applications-prod-5-southern'-array. With that my subdirectory would be `applicat-output`, i.e., this is not the name I passed and the string was truncated without any apparent reason. Due to this issue, my main suggestion is to constrain the choices of `dir_type` to a selection of allowed names, as I mentioned before.

File: tests/unit_tests/test_io_handler.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply tests/unit_tests/test_io_handler.py:38]: I meant that it would be good if every time `io_handler.use_plain_output_path` is True and the user passes `label` that the user is informed about the argument passed in `label` not being taken into account. But your argument above that the io_handler is not going to be used directly by external users (only by developers) convinced me that we might indeed not need the warning.

File: simtools/io_handler.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/io_handler.py:92]: I think I responded the first point above

**Comment by VictorBarbosaMartins** [comment]
Hi @GernotMaier , thanks a lot for the carefully reading and answering my comments. It is clearer to me now, and I understand that many of the comments are not directly related to the PR. I usually get a lot of those in my PRs as well, I thought it was fine. But I agree with you that it is probably better to focus on the PR and be more effective/faster.

I just have two remaining points, both with the dir_type parameter (see in the comments above), I am fine with the rest. Do you want to add something else @orelgueta ?

(1h02 min more for the re-review here)

File: simtools/io_handler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (orelgueta) [reply simtools/io_handler.py]: But if it is an example (which is what I read when I see `e.g.,`) then we shouldn't limit it. I don't think we should limit it here from the point of view of the code either.

File: simtools/io_handler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (orelgueta) [reply simtools/io_handler.py]: And I would add that we don't have to constrain, but just remove `simtools-` using regex if this is specifically your original goal here @GernotMaier (I must admit I couldn't figure out the reasoning behind it yesterday).

[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py]: You have a suggestion on the regex expression?

**Comment by orelgueta** [comment]
I still think that 
https://github.com/gammasim/simtools/blob/ffdc21416fba1602e030ed5123fe82eba5b14bea/simtools/io_handler.py#L92

is prone to errors unless you enforce `dir_type` to be exactly one of those three options `'simtools', 'test', 'simtools-result'`. As a developer, if I just see the documentation with "e.g.," then I think I can provide anything. 
A real world example would be `corsika-data` or `simtel-data`, for latter which I will get an output directory `simtel-d`. 

So I see two options for this specific PR
- You enforce only those three options for now and in the future we deal with other options when they come (probably me cause I will need to implement the corsika/simtel-data case).
- You use regex to remove `-result` at the end of `dir` type and add a mention of that in the doscrting. The actual command would be `re.sub(r"\-result$", "", dir_type)` (I would vote for this option).

File: simtools/io_handler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (orelgueta) [reply simtools/io_handler.py]: Given below (with another optional solution). 
And BTW, reading my comment again made me realise that even here I still did not understand what you wanted to do until I had to do it myself.

[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/io_handler.py]: OK - now I get what the issue is. Clearly missed it, as I was testing with existing use cases only. You need to make comments in review clear enough, I obviously would have done it differently if I would have been aware of it.

@orelgueta - thanks, I've implemented the solution.

I think we shouldn't fix the options of dir_type, unless there is a logical reason for it.

**Comment by VictorBarbosaMartins** [comment]
> 

Good that we agree in this issue @orelgueta . I think we have to understand better the reason behind this line. Why do we want in the first place to cut out '-results' whenever `dir_type` is passed as 'simtools-results' (and allow all the remaining possibilities). I dont mind the regex but I still prefer the solution I mentioned before, i.e., instead of passing `simtools-results` in the argument, use the `use_plain_output_path` attribute to have the same effect.

File: simtools/io_handler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (orelgueta) [reply simtools/io_handler.py]: OK, so if you implemented the regex solution, please add also a mention in the docstring.

**Comment by GernotMaier** [comment]
As mentioned several time: this is to ensure that the tools behave exactly the same way as before if you don't give output_path and use_plain_output_path as options. I think this is a good motivation.

We can discuss (beyond this pull request) if we should give up the old-style directory naming completely.  This pull request was about adding functionality, not to remove or change existing beavior.

**Comment by GernotMaier** [comment]
Tests are successful now, and I think I've addressed your comments (or we discussed them sufficiently).

Let me know if there is anything remaining to fix.

**Comment by orelgueta** [comment]
I didn't actually do a review in the end except that specific point so I let @VictorBarbosaMartins approve if he agrees.

**Comment by GernotMaier** [comment]
Added one more unit test for `dir_type=None`. This should be the last commit.

**Review by VictorBarbosaMartins — APPROVED** [review]
we have discussed enough, thank you for the improvements. please go ahead
